sling/nlp/ner/annotators.cc

```diff
@@ -0,0 +1,1062 @@
+    bool all_digits = true;
+    for (char c : word) {
+      if (c >= '0' && c <= '9') {
+        has_digits = true;
+      } else {
+        all_digits = true;
+      }
+    }
+    if (!has_digits) continue;
+
+    // Try to parse token as a number.
+    Handle number = ParseNumber(word, format);
+    if (!number.IsNil()) {
+      // Numbers between 1582 and 2038 are considered years.
+      int flags = SPAN_NUMBER;
+      if (word.size() == 4 && all_digits && number.IsInt()) {
+        int value = number.AsInt();
+        if (value >= 1582 && value <= 2038) {
+          Builder b(document->store());
+          b.AddIsA(n_time_);
+          b.AddIs(number);
+          number = b.Create().handle();
+          flags = SPAN_DATE;
+        }
+      }
+      chart->Add(t, t + 1, number, flags);
+    }
+  }
+}
+
+Handle NumberAnnotator::ParseNumber(Text str, char tsep, char dsep, char msep) {
+  const char *p = str.data();
+  const char *end = p + str.size();
+  if (p == end) return Handle::nil();
+
+  // Parse sign.
+  double scale = 1.0;
+  if (*p == '-') {
+    scale = -1.0;
+    p++;
+  } else if (*p == '+') {
+    p++;
+  }
+
+  // Parse integer part.
+  double value = 0.0;
+  const char *group = nullptr;
+  while (p < end) {
+    if (*p >= '0' && *p <= '9') {
+      value = value * 10.0 + (*p++ - '0');
+    } else if (*p == tsep) {
+      if (group != nullptr && p - group != 3) return Handle::nil();
+      group = p + 1;
+      p++;
+    } else if (*p == dsep) {
+      break;
+    } else {
+      return Handle::nil();
+    }
+  }
+  if (group != nullptr && p - group != 3) return Handle::nil();
+
+  // Parse decimal part.
+  bool decimal = false;
+  if (p < end && *p == dsep) {
+    decimal = true;
+    p++;
+    group = nullptr;
+    while (p < end) {
+      if (*p >= '0' && *p <= '9') {
+        value = value * 10.0 + (*p++ - '0');
+        scale /= 10.0;
+      } else if (*p == msep) {
+        if (group != nullptr && p - group != 3) return Handle::nil();
+        group = p + 1;
+        p++;
+      } else {
+        return Handle::nil();
+      }
+    }
+    if (group != nullptr && p - group != 3) return Handle::nil();
+  }
+  if (p != end) return Handle::nil();
+
+  // Compute number.
+  value *= scale;
+  if (decimal || value < Handle::kMinInt || value > Handle::kMaxInt) {
+    return Handle::Float(value);
+  } else {
+    return Handle::Integer(value);
+  }
+}
+
+Handle NumberAnnotator::ParseNumber(Text str, Format format) {
+  Handle number = Handle::nil();
+  switch (format) {
+    case STANDARD:
+      number = ParseNumber(str, '.', ',', 0);
+      if (number.IsNil()) number = ParseNumber(str, ',', '.', 0);
+      break;
+    case IMPERIAL:
+      number = ParseNumber(str, ',', '.', 0);
+      if (number.IsNil()) number = ParseNumber(str, '.', ',', 0);
+      break;
+    case NORWEGIAN:
+      number = ParseNumber(str, ' ', '.', ' ');
+      if (number.IsNil()) number = ParseNumber(str, '.', ',', 0);
+      break;
```

Comment:
The number annotator tries to deal with the ambiguity between differences in decimal and thousand separators by checking for the tree-digit groupings. If we allow the thousand groups to be two or three digits except for the last, it will not add any new ambiguities, as long as we check that the last group is three digits. I have also added support for "unit of amounts" (Q2360980)in the NumberScaleAnnotator so we can parse things like "5.6 lakhs" correctly:

>So [5.6 lakhs|{=#1 :/w/quantity /w/amount: 560000}] would be written as [5,60,000|{=#2 :/w/quantity /w/amount: 560000}] meaning [560,000|{=#3 :/w/quantity /w/amount: 560000}], and [2.3 crores|{=#4 :/w/quantity /w/amount: 2.3e+07}] would be [2,30,00,000|{=#5 :/w/quantity /w/amount: 23000000}] meaning [23,000,000|{=#6 :/w/quantity /w/amount: 23000000}].